### PR TITLE
feat: setFocus on dismiss button method  #740 

### PR DESCRIPTION
**Pull request by driskull**
**Related Issue:** #740 

## Summary

feat: #740 setFocus on dismiss button method

### CODE REVIEW

File: src/components/calcite-panel/calcite-panel.tsx [hunk 1]
```diff
@@ -102,6 +114,17 @@ export class CalcitePanel {
...
   };
 
+  // --------------------------------------------------------------------------
+  //
+  //  Methods
+  //
+  // --------------------------------------------------------------------------
+
+  @Method()
+  async setFocusDismissButton() {
```
Comment (pr3tori4n) [inline src/components/calcite-panel/calcite-panel.tsx]: maybe just call it `focusDismissButton`

File: src/demos/panel/basic.html [hunk 2]
```diff
@@ -53,6 +53,9 @@ <h2>Dismissible Panel</h2>
           dismissiblePanel.addEventListener("calcitePanelDismissedChange", function() {
             console.log("calcitePanelDismiss event");
           });
+          setTimeout(function() {
+            dismissiblePanel.setDismissFocus();
```
Comment (pr3tori4n) [inline src/demos/panel/basic.html]: method name doesn't match.

File: src/components/calcite-panel/calcite-panel.tsx [hunk 3]
```diff
@@ -102,6 +114,17 @@ export class CalcitePanel {
...
   };
 
+  // --------------------------------------------------------------------------
+  //
+  //  Methods
+  //
+  // --------------------------------------------------------------------------
+
+  @Method()
+  async focusDismissButton() {
```
Comment (jcfranco) [inline src/components/calcite-panel/calcite-panel.tsx]: Code looks good, but I had a general question. Are we OK with having individual focus methods for component parts? This does set a precedent for that. 

Just as an idea, we could use `setFocus()` for the default focus behavior and allow an argument for a specific part.

```ts
component.setFocus(); // default focus
component.setFocus("some-part"); // specialized focus
```

This does become tricky when there is only one part and no default focus behavior (like in calcite-panel 😅).

Comment (jcfranco) [inline src/components/calcite-panel/calcite-panel.tsx]: The name is now inconsistent w/ `calcite-action`: https://github.com/Esri/calcite-app-components/blob/9f02abe53f0ce530b90a2b8a5adeb1a290653085/src/components/calcite-action/calcite-action.tsx#L90

Comment (driskull) [reply src/components/calcite-panel/calcite-panel.tsx]: I like this idea. even though it is a little tricky.

File: src/components/calcite-panel/calcite-panel.e2e.ts [hunk 4]
```diff
@@ -52,15 +52,15 @@ describe("calcite-panel", () => {
...
       withPeerDependencies: true
     });
 
     const panel = await page.find("calcite-panel");
 
-    const focusDismissButton = panel.callMethod("focusDismissButton");
+    const setFocus = panel.callMethod("setFocus");
 
-    expect(focusDismissButton).toBeInstanceOf(Promise);
+    expect(setFocus).toBeInstanceOf(Promise);
```
Comment (jcfranco) [inline src/components/calcite-panel/calcite-panel.e2e.ts]: I think this also needs to assert on the final focused element.

Comment (driskull) [reply src/components/calcite-panel/calcite-panel.e2e.ts]: I don't think focus works very well in e2e tests. Tried it but couldn't get it to work.

File: src/components/calcite-panel/calcite-panel.tsx [hunk 5]
```diff
@@ -102,6 +116,19 @@ export class CalcitePanel {
...
 
+  // --------------------------------------------------------------------------
+  //
+  //  Methods
+  //
+  // --------------------------------------------------------------------------
+
+  @Method()
+  async setFocus(focusId?: FocusId) {
+    if (focusId === "dismiss-button") {
```
Comment (pr3tori4n) [inline src/components/calcite-panel/calcite-panel.tsx]: not sure the purpose of this argument. If the only valid value is dismiss-button, why not just have it hardedcoded to target that element and remove the argument altogether?

File: src/components/calcite-panel/calcite-panel.e2e.ts (hunk 4 above)
Comment (pr3tori4n) [reply src/components/calcite-panel/calcite-panel.e2e.ts]: maybe try 
```const someId = page.evaluate( () => {
const panel = document.querySelector("calcite-panel");
  const activeElement = panel.shadowRoot.activeElement;
  return activeElement.id; //or tagname or some attribute you can use to validate its the right one.
});

File: src/components/calcite-panel/calcite-panel.tsx (hunk 5 above)
Comment (driskull) [reply src/components/calcite-panel/calcite-panel.tsx]: Because we may have the ability to focus a panel in the future.

Comment (driskull) [reply src/components/calcite-panel/calcite-panel.tsx]: Update: added ability to focus on container.

**Pull request merged**